@xen-orchestra/fs/src/azure.js

```diff
@@ -0,0 +1,195 @@
+  #container
+  #dir
+  #blobServiceClient
+  #containerClient
+
+  constructor(remote, _opts) {
+    super(remote)
+    const { username, path, password, host, protocol } = parse(remote.url)
+    /* 
+For Azure not Azurite
+ --------------------
+  const credentials = new StorageSharedKeyCredential(username, password);
+  this.#blobServiceClient = new BlobServiceClient(`${protocol}://${username}.blob.core.windows.net`, credentials);
+ --------------------
+*/
+
+    // this works ONLY for Azurite
+    this.#blobServiceClient = BlobServiceClient.fromConnectionString(
+      `DefaultEndpointsProtocol=${protocol};AccountName=${username};AccountKey=${password};BlobEndpoint=${protocol}://${host}/${username}`
+    )
+
+    const parts = split(path)
+    this.#container = parts.shift()
+    this.#dir = join(...parts)
+    this.#containerClient = this.#blobServiceClient.getContainerClient(this.#container)
+    this.#createContainer()
+
+    const WITH_RETRY = ['_copy', '_getSize', '_list', '_outputFile', '_read', '_rename', '_unlink', '_writeFile']
+    WITH_RETRY.forEach(functionName => {
+      if (this[functionName] !== undefined) {
+        this[functionName] = pRetry.wrap(this[functionName], {
+          delays: [100, 200, 500, 1000, 2000],
+          when: err => !['EEXIST', 'EISDIR', 'ENOTEMPTY', 'ENOENT', 'ENOTDIR', 'EISDIR'].includes(err?.code),
+          onRetry(error) {
+            warn('retrying method on fs ', {
+              method: functionName,
+              attemptNumber: this.attemptNumber,
+              delay: this.delay,
+              error,
+              file: this.arguments?.[0],
+            })
+          },
+        })
+      }
+    })
+  }
+
+  get type() {
+    return 'azure'
+  }
+
+  async #createContainer() {
+    await this.#containerClient.createIfNotExists()
+  }
+
+  async #makePrefix(path) {
+    return path.endsWith('/') ? path : `${path}/`
+  }
+
+  async #isNotEmptyDir(dir) {
+    const prefix = this.#makePrefix(dir)
+    const blobs = []
+    for await (const blob of this.#containerClient.listBlobsFlat({ prefix })) {
+      blobs.push(blob.name)
+    }
+
+    return blobs.length > 0
+  }
+
+  async #streamToBuffer(readableStream, buffer) {
+    return new Promise((resolve, reject) => {
+      let bytesRead = 0
+
+      readableStream.on('data', data => {
+        if (bytesRead + data.length <= buffer.length) {
+          data.copy(buffer, bytesRead)
+          bytesRead += data.length
+        } else {
+          reject(new Error('Buffer size exceeded'))
+        }
+      })
+      readableStream.on('end', () => {
+        resolve(bytesRead)
+      })
+      readableStream.on('error', reject)
+    })
+  }
+
+  // this can be used right away instead of _writeFile func
+  async #_outputStream(file, data) {
+    const blobClient = this.#containerClient.getBlockBlobClient(file)
+    await blobClient.uploadStream(data, MAX_BLOCK_SIZE, MAX_BLOCK_COUNT)
+  }
```

Comment:
please test if this store the block in memory 

if yes, compute the minimal block size depending n the stream expected size